File: FinalProject_OOP_Java/src/Vehicle/Vehicle_Electric.java
package Vehicle;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.text.ParseException;
import java.util.Date;

public class Vehicle_Electric extends Vehicle implements Electricity_Vehicle
{
	 //Battery là điện
    //dịch vụ khác là sạc điện
	private double Veh_dTime_Charge; //Thời gian sạc
    private double Veh_dElec_Capacity; //Dung lượng bình sạc
	

	public double getVeh_dTime_Charge() {
		return Veh_dTime_Charge;
	}

	public void setVeh_dTime_Charge(double veh_dTime_Charge) {
		Veh_dTime_Charge = veh_dTime_Charge;
	}

	public double getVeh_dElec_Capacity() {
		return Veh_dElec_Capacity;
	}

	public void setVeh_dElec_Capacity(double veh_dElec_Capacity) {
		if(veh_dElec_Capacity <= 0)
			throw new ArithmeticException("EROLL");
		Veh_dElec_Capacity = veh_dElec_Capacity;
	}

	public Vehicle_Electric()
    {
		super();
        this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String type, String locate, Date in_time, Date out_time, int battery, double capacity) 
    {
		super(iD, type, locate, in_time, out_time);
        this.setVeh_iBattery(battery);
        this.setVeh_dElec_Capacity(capacity);
        this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String type, String locate, Date in_time, Date out_time, int battery) 
    {
		super(iD, type, locate, in_time, out_time);
		this.setVeh_iBattery(battery);
		this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String type, String locate, Date in_time, int battery, double capacity) 
    {
		super(iD, type, locate, in_time);
        this.setVeh_iBattery(battery);
        this.setVeh_dElec_Capacity(capacity);
        this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String locate, Date in_time, Date out_time, int battery, double capacity) 
    {
		super(iD, locate, in_time);
        this.setVeh_iBattery(battery);
        this.setVeh_dElec_Capacity(capacity);
        this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String locate, Date in_time, double capacity) 
    {
		super(iD, locate, in_time);
        this.setVeh_dElec_Capacity(capacity);
        this.setVeh_iMoney(70);
    }
	public Vehicle_Electric(String iD, String type, String locate, Date in_time, int battery) 
    {
		super(iD, type, locate, in_time);
        this.setVeh_iBattery(battery);
        this.setVeh_iMoney(70);
    }
	 public Vehicle_Electric(String iD, Date in_time, Date out_time, double capacity)  
     {
		 super(iD,in_time,out_time);
         this.setVeh_dElec_Capacity(capacity);
         this.setVeh_iMoney(70);
     }
	 public Vehicle_Electric(String iD, Date in_time, double capacity) 
     {
		 super(iD,in_time);
         this.setVeh_dElec_Capacity(capacity);
         this.setVeh_iMoney(70);
     }
	 
	 public void Veh_Input_OutTime() throws IOException, ParseException
     {
         super.Veh_Input_OutTime();
     }
	 
	 public void Veh_Input_OutTime(String out_time) throws ParseException
     {
         super.Veh_Input_OutTime(out_time);
     }
	 
	 public void Veh_Input(String iD, String type, String locate, Date in_time, Date out_time, int battery, double capacity)
     {
         super.Veh_Input(iD, type, locate, in_time, out_time);
         this.setVeh_iBattery(battery);
         this.setVeh_dElec_Capacity(capacity);
     }
	 
	@Override
	public void Veh_Input_Other_services() throws NumberFormatException, IOException {
		// TODO Auto-generated method stub
		System.out.println("Sac pin hay khong?");
        System.out.println("1.Sac");
        System.out.println("2.Khong sac");
        InputStreamReader ips= new InputStreamReader(System.in);
        BufferedReader br =new BufferedReader(ips);
        this.Veh_iOther_services = Integer.parseInt(br.readLine());
	}
	
	@Override
	public void Veh_Input_Other_services(int x) {
		// TODO Auto-generated method stub
		this.setVeh_iOther_services(x);
	}
	
	public void Veh_Output()
    {
        super.Veh_Output();
        System.out.printf("%10s|\n", " Electric");
    }

	public long Veh_Calculate_Time()
    {
        return super.Veh_Calculate_Time();
    }
	
	@Override
	public double Veh_Calculate_Money_Parking() {
		// TODO Auto-generated method stub
		long timeMin = this.Veh_Calculate_Time();
        return this.Veh_iMoney * timeMin;
	}
	
	public void Information()
    {
        System.out.println("Thoi gian sac pin:" + this.Veh_dTime_Charge);
        System.out.println("Tien gui xe:" + this.Veh_Calculate_Money_Parking());
        System.out.println("Tien dich vu:" + this.Veh_Calculate_Money_Other_services());
        System.out.println("Tong:" + this.Veh_Calculate_Money_Sum());
    }

	@Override
	public double Veh_Calculate_Money_Other_services() {
		// TODO Auto-generated method stub
		double tong = 0;
        if (this.Veh_iOther_services == 1)
        {
        	this.Veh_Calculate_Time_Charge(20000);
            double time = this.Veh_Calculate_Time();
            double Hours = time / 60 ;
            double Hours_Charge = this.Veh_dTime_Charge;
            double x = Hours_Charge - Hours;
            if (x < 0)
                tong += 1700 * Hours_Charge;
            else
                tong += 1700 * Hours;
        }

        return tong;
	}
	
	public double Veh_Calculate_Money_Sum()
    {
        return super.Veh_Calculate_Money_Sum();
    }

    //hàm tính thời gian sạc // (dung luong, P) tg=dl/cs
    public void Veh_Calculate_Time_Charge(double Wattage)
    {
        double t = Math.round(this.Veh_dElec_Capacity / Wattage*100)/100;
        this.Veh_dTime_Charge = t;
    }
    
    // string đường đi
    public String path_Vehicle()
    {
        return super.path_Vehicle();
    }
    
    public String toString()
    {
        return super.toString() + this.Veh_dElec_Capacity +";" + this.Veh_iOther_services;
    }
}

File: FinalProject_OOP_C#/Vehicle_Electric.cs
using Microsoft.VisualBasic.FileIO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Metadata.Ecma335;
using System.Text;
using System.Threading.Tasks;

namespace FinalProject_OOP
{
    class Vehicle_Electric : Vehicle, Electricity_Vehicle
    {
        //Battery là điện
        //dịch vụ khác là sạc điện
        private TimeSpan Veh_dTime_Charge; //Thời gian sạc
        private double Veh_dElec_Capacity; //Dung lượng bình sạc
        public TimeSpan Veh_Time_Charge
        {
            get { return this.Veh_dTime_Charge; }
            set
            {
                this.Veh_dTime_Charge = value;
            }
        }

        public double Veh_Elec_Capacity
        {
            get { return this.Veh_dElec_Capacity; }
            set
            {
                if (value <= 0)
                    throw new ArgumentOutOfRangeException(
                         $"{nameof(value)} khong hop le !");
                this.Veh_dElec_Capacity = value;
            }
        }

        public Vehicle_Electric() : base()
        {
            this.Veh_dTime_Charge = new TimeSpan();
            this.Veh_Money = 70;
        }

        public Vehicle_Electric(string iD, string type, string locate, DateTime in_time, DateTime out_time, int battery, double capacity) : base(iD, type, locate, in_time, out_time)
        {
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
        }

        public Vehicle_Electric(string iD, string type, string locate, DateTime in_time, DateTime out_time, int battery) : base(iD, type, locate, in_time, out_time)
        {
            this.Veh_Battery = battery;
            this.Veh_Money = 70;
        }
        public Vehicle_Electric(string iD, string type, string locate, DateTime in_time, int battery, double capacity) : base(iD, type, locate, in_time)
        {
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
            this.Veh_Out_time=DateTime.Now;
        }

        public Vehicle_Electric(string iD, string locate, DateTime in_time, DateTime out_time, int battery, double capacity) : base(iD, locate, in_time)
        {
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
        }

        public Vehicle_Electric(string iD, string locate, DateTime in_time, double capacity) : base(iD, locate, in_time)
        {
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
        }

        public Vehicle_Electric(string iD, string type, string locate, DateTime in_time, int battery) : base(iD, type, locate, in_time)
        {
            this.Veh_Battery = battery;
            this.Veh_Money = 70;
        }
        public Vehicle_Electric(string iD, DateTime in_time, DateTime out_time, double capacity) : base(iD, in_time, out_time)
        {
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
        }

        public Vehicle_Electric(string iD, DateTime in_time, double capacity) : base(iD, in_time)
        {
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Money = 70;
        }
        ~Vehicle_Electric()
        {
        }

        public override void Veh_Input()
        {
            base.Veh_Input();
            Console.Write("Nhap phan tram pin: ");
            this.Veh_Battery = int.Parse(Console.ReadLine());
            Console.WriteLine("Nhap Capacity: ");
            this.Veh_Elec_Capacity = double.Parse(Console.ReadLine());
        }

        public override void Veh_Input_OutTime()
        {
            base.Veh_Input_OutTime();
        }
        public override void Veh_Input_OutTime(string out_time)
        {
            base.Veh_Input_OutTime(out_time);
        }
        public void Veh_Input(string iD, string type, string locate, DateTime in_time, DateTime out_time, int battery, double capacity)
        {
            base.Veh_Input(iD, type, locate, in_time, out_time);
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
        }

        public void Veh_Input(string iD, string type, string locate, DateTime in_time, int battery, double capacity,int OtherService)
        {
            base.Veh_Input(iD, type, locate, in_time);
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
            this.Veh_Other_services = OtherService;
        }

        public void Veh_Input(string iD, string locate, DateTime in_time, DateTime out_time, int battery, double capacity)
        {
            base.Veh_Input(iD, locate, in_time);
            this.Veh_Battery = battery;
            this.Veh_Elec_Capacity = capacity;
        }

        public void Veh_Input(string iD, string locate, DateTime in_time, double capacity)
        {
            base.Veh_Input(iD, locate, in_time);
            this.Veh_Elec_Capacity = capacity;
        }

        public void Veh_Input(string iD, string type, string locate, DateTime in_time, int battery,int service)
        {
            base.Veh_Input(iD, type, locate, in_time);
            this.Veh_Battery = battery;
            this.Veh_Other_services=service;
        }
        public void Veh_Input(string iD, DateTime in_time, DateTime out_time, double capacity)
        {
            base.Veh_Input(iD, in_time, out_time);
            this.Veh_Elec_Capacity = capacity;
        }

        public void Veh_Input(string iD, DateTime in_time, double capacity)
        {
            base.Veh_Input(iD, in_time);
            this.Veh_Elec_Capacity = capacity;
        }

        public override void Veh_Input_Other_services()
        {
            Console.WriteLine("Sac pin hay khong?");
            Console.WriteLine("1.Sac");
            Console.WriteLine("2.Khong sac");
            this.Veh_Other_services = int.Parse(Console.ReadLine());
        }

        public override void Veh_Input_Other_services(int x)
        {
            this.Veh_Other_services = x;
        }

        public override void Veh_Output()
        {
            base.Veh_Output();
            Console.WriteLine("{0,10}|"," Electric");
        }

        public override TimeSpan Veh_Calculate_Time()
        {
            return base.Veh_Calculate_Time();
        }

        public override double Veh_Calculate_Money_Parking()
        {
            TimeSpan time = this.Veh_Calculate_Time();
            double Minutes = time.Days * 24 * 60 + time.Hours * 60 + time.Minutes + time.Seconds / 60;
            return this.Veh_Money * Minutes;
        }
        public void Information()
        {
            Console.WriteLine("Thoi gian sac pin:" + this.Veh_Time_Charge.ToString());
            Console.WriteLine("Tien gui xe:" + this.Veh_Calculate_Money_Parking());
            Console.WriteLine("Tien dich vu:" + this.Veh_Calculate_Money_Other_services());
            Console.WriteLine("Tong:" + this.Veh_Calculate_Money_Sum());
        }
        public override double Veh_Calculate_Money_Other_services()
        {

            double tong = 0;
            if (this.Veh_Other_services == 1)
            {
                TimeSpan time = this.Veh_Calculate_Time();
                double Hours = time.Days * 24 + time.Hours + time.Minutes / 60 + time.Seconds / (60 * 60);
                double Hours_Charge = this.Veh_Time_Charge.Days * 24 + this.Veh_Time_Charge.Hours + this.Veh_Time_Charge.Minutes / 60 + this.Veh_Time_Charge.Seconds / (60 * 60);
                double x = Hours_Charge - Hours;
                if (x < 0)
                    tong += 1700 * Hours_Charge;
                else
                    tong += 1700 * Hours;
            }

            return tong;
        }
        // Tinh tien tong
        public override double Veh_Calculate_Money_Sum()
        {
            return base.Veh_Calculate_Money_Sum();
        }

        //hàm tính thời gian sạc // (dung luong, P) tg=dl/cs
        public void Veh_Calculate_Time_Charge(double Wattage)
        {
            double t = Math.Round(this.Veh_dElec_Capacity / Wattage, 2);
            int hours = (int)t;
            int minutes = (int)((t - hours) * 60);
            int seconds = (int)((t - minutes) * 60);
            this.Veh_Time_Charge = new TimeSpan(hours, minutes, seconds);
        }

        // string đường đi
        public override string path_Vehicle()
        {
            return base.path_Vehicle();
        }

        public static explicit operator Vehicle_Bike(Vehicle_Electric x)
        {
            Vehicle_Bike kq = new Vehicle_Bike();
            kq.Veh_ID = x.Veh_ID;
            kq.Veh_Type = x.Veh_Type;
            kq.Veh_Locate = x.Veh_Locate;
            kq.Veh_In_time = x.Veh_In_time;
            kq.Veh_Out_time = x.Veh_Out_time;
            kq.Veh_Battery = x.Veh_Battery; // phan tram pin
            return kq;
        }

        public static explicit operator Vehicle_Car(Vehicle_Electric x)
        {
            Vehicle_Car kq = new Vehicle_Car();
            kq.Veh_ID = x.Veh_ID;
            kq.Veh_Type = x.Veh_Type;
            kq.Veh_Locate = x.Veh_Locate;
            kq.Veh_In_time = x.Veh_In_time;
            kq.Veh_Out_time = x.Veh_Out_time;
            kq.Veh_Battery = x.Veh_Battery;
            return kq;
        }

        public static explicit operator Vehicle_Motobike(Vehicle_Electric x)
        {
            Vehicle_Motobike kq = new Vehicle_Motobike();
            kq.Veh_ID = x.Veh_ID;
            kq.Veh_Type = x.Veh_Type;
            kq.Veh_Locate = x.Veh_Locate;
            kq.Veh_In_time = x.Veh_In_time;
            kq.Veh_Out_time = x.Veh_Out_time;
            kq.Veh_Battery = x.Veh_Battery;
            return kq;
        }

        public static explicit operator Vehicle_Scooter(Vehicle_Electric x)
        {
            Vehicle_Scooter kq = new Vehicle_Scooter();
            kq.Veh_ID = x.Veh_ID;
            kq.Veh_Type = x.Veh_Type;
            kq.Veh_Locate = x.Veh_Locate;
            kq.Veh_In_time = x.Veh_In_time;
            kq.Veh_Out_time = x.Veh_Out_time;
            kq.Veh_Battery = x.Veh_Battery;
            return kq;
        }

        public static explicit operator Vehicle_Truck(Vehicle_Electric x)
        {
            Vehicle_Truck kq = new Vehicle_Truck();
            kq.Veh_ID = x.Veh_ID;
            kq.Veh_Type = x.Veh_Type;
            kq.Veh_Locate = x.Veh_Locate;
            kq.Veh_In_time = x.Veh_In_time;
            kq.Veh_Out_time = x.Veh_Out_time;
            kq.Veh_Battery = x.Veh_Battery;
            return kq;
        }

        public override string toString()
        {
            return base.toString() + this.Veh_Elec_Capacity +";" + this.Veh_Other_services;
        }
    }
}
